vortex-dtype/src/struct_.rs

```diff
@@ -210,60 +198,63 @@ impl StructDType {
         &self.names
     }
 
-    /// Find the index of a field.
-    pub fn find(&self, field: &Field) -> Option<usize> {
-        match field {
-            Field::Name(name) => self.find_name(name),
-            Field::Index(idx) => Some(*idx),
-        }
+    /// Returns the number of fields in the struct
+    pub fn nfields(&self) -> usize {
+        self.names.len()
+    }
+
+    /// Returns the name of the field at the given index
+    pub fn field_name(&self, index: usize) -> VortexResult<&FieldName> {
+        self.names
+            .get(index)
+            .ok_or_else(|| vortex_err!("field index out of bounds"))
     }
 
     /// Find the index of a field by name
     /// Returns `None` if the field is not found
-    pub fn find_name(&self, name: &str) -> Option<usize> {
-        self.names.iter().position(|n| n.as_ref() == name)
+    pub fn find(&self, name: impl AsRef<str>) -> VortexResult<usize> {
+        let name = name.as_ref();
+        self.names
+            .iter()
+            .position(|n| n.as_ref() == name)
+            .ok_or_else(|| {
+                vortex_err!(
+                    "Field {} not found in {}",
+                    name,
+                    self.names.iter().join(", ")
+                )
+            })
     }
 
-    /// Get information about the referenced field, either by name or index
-    /// Returns an error if the field is not found
-    pub fn field_info(&self, field: &Field) -> VortexResult<FieldInfo> {
-        let index = match field {
-            Field::Name(name) => self
-                .find_name(name)
-                .ok_or_else(|| vortex_err!("Unknown field: {name}"))?,
-            Field::Index(index) => *index,
-        };
-        if index >= self.names.len() {
-            vortex_bail!("field index out of bounds: {index}")
-        }
-        Ok(FieldInfo {
-            index,
-            name: self.names[index].clone(),
-            dtype: self.dtypes[index].clone(),
-        })
+    /// Get the [`DType`] of a field.
+    pub fn field(&self, name: impl AsRef<str>) -> VortexResult<DType> {
```

Comment:
this will eagerly materialize the underlying child, don't think we want this

Reply:
That's fine, it will only go one level down, that's kind of the point?

Reply:
you're right, too many layers of abstraction